In [27]:
import threading
import time

start=time.perf_counter() # simple function

def do_something():
    print(f'sleeping')
    time.sleep(1)
    print("Done something...")

t1=threading.Thread(target=do_something) # creating a thread
t1.start() #thread execution starts 
t1.join() # thread execution get completed
finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds') # printing the thread execution time

sleeping
Done something...
Finished in 1.0 seconds


In [26]:
import threading
import time

start=time.perf_counter()

def do_something(): # simple functions
    print(f'sleeping')
    time.sleep(1)
    print("Done something...")

threads=[]

for _ in range(50):
    t=threading.Thread(target=do_something)
    t.start()
    threads.append(t)
    
for thread in threads:
    thread.join()

finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds')# printing the thread execution time

sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
sleeping
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done somet

In [22]:
import threading
import time

start=time.perf_counter()

def do_something(seconds): # simple functions with argument
    print(f'sleeping {seconds} second(s)')
    time.sleep(seconds)
    print("Done something...")

threads=[]

for _ in range(10):
    t=threading.Thread(target=do_something,args=[1.2])
    t.start()
    threads.append(t)
    
for thread in threads:
    thread.join()

finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds')# printing the thread execution time

sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
sleeping 1.2 second(s)
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Done something...
Finished in 1.2 seconds


In [23]:
import concurrent.futures
import time

start=time.perf_counter()

def do_something(seconds): # simple functions with argument
    print(f'sleeping {seconds} second(s)')
    time.sleep(seconds) #sleep function of thread 
    return 'Done something...'

#using concurent.futures module 
#Using ThreadPoolExecutor 

with concurrent.futures.ThreadPoolExecutor() as exe:
    f1=exe.submit(do_something,1.5) # submit() function stores the future object
    print(f1.result())

finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds')# printing the thread execution time


sleeping 1.5 second(s)
Done something...
Finished in 1.5 seconds


In [21]:
import concurrent.futures
import time

start=time.perf_counter()

def do_something(seconds): # simple functions with argument
    print(f'sleeping {seconds} second(s)')
    time.sleep(seconds) #sleep function of thread 
    return 'Done something...'

#using concurent.futures module 
#Using ThreadPoolExecutor 

with concurrent.futures.ThreadPoolExecutor() as exe:
    
    res=[exe.submit(do_something,1.5) for _ in range(2)]# creating mutiple threads using list comprehension
    for f in concurrent.futures.as_completed(res):  # concurrent.futures.as_completed() takes  list of future object and returns iterators 
         print(f.result())

finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds') # printing the thread execution time


sleeping 1.5 second(s)
sleeping 1.5 second(s)
Done something...
Done something...
Finished in 1.5 seconds


In [30]:
# take series of arguments 
import concurrent.futures
import time

start=time.perf_counter()

def do_something(seconds): # simple functions with argument
    print(f'sleeping {seconds} second(s)')
    time.sleep(seconds) #sleep function of thread 
    return 'Done something...'

#using concurent.futures module 
#Using ThreadPoolExecutor 

with concurrent.futures.ThreadPoolExecutor() as exe:
    secs=[5,4,3,2,1] # series of arguments 
    res=[exe.submit(do_something,sec) for sec in secs]# creating mutiple threads using list comprehension
    for f in concurrent.futures.as_completed(res):  # concurrent.futures.as_completed() takes  list of future object and returns iterators 
         print(f.result())

finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds') # printing the thread execution time


sleeping 5 second(s)
sleeping 4 second(s)
sleeping 3 second(s)
sleeping 2 second(s)
sleeping 1 second(s)
Done something...
Done something...
Done something...
Done something...
Done something...
Finished in 5.0 seconds


In [32]:
# Using map function
import concurrent.futures
import time

start=time.perf_counter()

def do_something(seconds): # simple functions with argument
    print(f'sleeping {seconds} second(s)')
    time.sleep(seconds) #sleep function of thread 
    return 'Done something...'

#using concurent.futures module 
#Using ThreadPoolExecutor 

with concurrent.futures.ThreadPoolExecutor() as exe:
    secs=[5,4,3,2,1] # series of arguments 
    res=exe.map(do_something,secs)# creating multiple threads using map function

for result in res:
        print(result) # no need concurrent.futures.as_completed() function since map function exactly maps the arguments to the function 
    # one-by -one.
    

finish=time.perf_counter()

print(f'Finished in {round(finish-start,2)} seconds') # printing the thread execution time


sleeping 5 second(s)
sleeping 4 second(s)
sleeping 3 second(s)
sleeping 2 second(s)
sleeping 1 second(s)
Done something...
Done something...
Done something...
Done something...
Done something...
Finished in 5.0 seconds


In [37]:
import time
import requests

audio_files=['https://commondatastorage.googleapis.com/codeskulptor-assets/sounddogs/missile.mp3',
            'https://codeskulptor-demos.commondatastorage.googleapis.com/descent/background%20music.mp3', 
            'https://pixabay.com/music/orchestral-cyberpunk-short-trailer-589196/', 
            'https://pixabay.com/music/alternative-hip-hop-battle-hip-hop-584304/',
            'https://pixabay.com/music/future-bass-that-future-bass-581055/'
]

t1=time.perf_counter()
for a_url in audio_files:
    a_bytes=requests.get(a_url).content
    a_name=a_url.split('/')[4]
    a_name=f'{a_name}.mp3'
    with open (a_name,'wb') as a_file:
        a_file.write(a_bytes)
        print(f'{a_name} was downloaded..')

t2=time.perf_counter()
print(f'Finished in {t2-t1} seconds')

sounddogs.mp3 was downloaded..
background%20music.mp3.mp3 was downloaded..
orchestral-cyberpunk-short-trailer-589196.mp3 was downloaded..
alternative-hip-hop-battle-hip-hop-584304.mp3 was downloaded..
future-bass-that-future-bass-581055.mp3 was downloaded..
Finished in 1.6085813979989325 seconds


In [38]:
import time
import requests
import concurrent.futures

audio_files=['https://commondatastorage.googleapis.com/codeskulptor-assets/sounddogs/missile.mp3',
            'https://codeskulptor-demos.commondatastorage.googleapis.com/descent/background%20music.mp3', 
            'https://pixabay.com/music/orchestral-cyberpunk-short-trailer-589196/', 
            'https://pixabay.com/music/alternative-hip-hop-battle-hip-hop-584304/',
            'https://pixabay.com/music/future-bass-that-future-bass-581055/'
]

t1=time.perf_counter()
def download_urls(audio_files):
    a_bytes=requests.get(a_url).content
    a_name=a_url.split('/')[4]
    a_name=f'{a_name}.mp3'
    with open (a_name,'wb') as a_file:
        a_file.write(a_bytes)
        print(f'{a_name} was downloaded..')

with concurrent.futures.ThreadPoolExecutor()as exe:
    exe.map(download_urls,audio_files)

t2=time.perf_counter()
print(f'Finished in {t2-t1} seconds')

future-bass-that-future-bass-581055.mp3 was downloaded..
future-bass-that-future-bass-581055.mp3 was downloaded..
future-bass-that-future-bass-581055.mp3 was downloaded..
future-bass-that-future-bass-581055.mp3 was downloaded..
future-bass-that-future-bass-581055.mp3 was downloaded..
Finished in 0.17674329000146827 seconds


I have 5 audio files which are about 5-10 mb each. It would be hard while downloading them manually one by one.

Using simple thread or simple function it take this much time to download: 1.6085813979989325 seconds

Whereas, Using concurrent.futures i.e; using multi-threading the amount of downloading the files reduced to 0.17674329000146827 seconds.